File: sw/application/application-toolkit/file-packer/FilePacker.py
#!/usr/bin/env python3
# *******************************************************************************
# Copyright (C) Altera Corporation
#
# This code and the related documents are Altera copyrighted materials and your
# use of them is governed by the express license under which they were provided to
# you ("License"). This code and the related documents are provided as is, with no
# express or implied warranties other than those that are expressly stated in the
# License.
# *******************************************************************************

import struct
import sys
from pathlib import Path

# To convert the output of this script to an object for linking:
# $LD -r -b binary -o WebServerHome.o WebServerHome.bin

class FileBlobHeader:
    def __bytes__(self):
        blob = bytearray()
        blob += struct.pack('<I', self._size)
        blob += struct.pack('<{}s'.format(FileBlobHeader._filenameSize), bytes(self._relativeFileName, 'utf-8'))
        return bytes(blob)

    def HeaderSize():
        return 4 + FileBlobHeader._filenameSize

    _filenameSize=128
    _size = 0
    _relativeFileName=""

if len(sys.argv) != 3:
    print("Usage: %s <source-folder> <output-file>\n".format(sys.argv[0]))
    sys.exit(-1)

fileBlobHeaderSize = FileBlobHeader.HeaderSize()

startFolderStr = sys.argv[1]
outputFilename = sys.argv[2]

pathlist = Path(startFolderStr).glob('**/*')

# Allocate a blob to contain all the files, with their headers
bigBlob = bytearray()

result = True
filenameSize = FileBlobHeader._filenameSize

supportedExtensions = ['.html', '.htm', '.css', '.js', '.png', '.ico', '.ttf']

for filePath in pathlist:
    if filePath.is_file():
        fileExtension = filePath.suffix
        if fileExtension.lower() not in supportedExtensions:
            continue

        pBlobHeader = FileBlobHeader()
        pBlobHeader._size = filePath.stat().st_size + FileBlobHeader.HeaderSize()

        # Compute relative path from start folder
        relativePath = filePath.relative_to(startFolderStr)

        # Copy the relative path name into the header
        pBlobHeader._relativeFileName = str(relativePath)
        print(pBlobHeader._relativeFileName)

        # Read each file into the blob
        try:
            with open(filePath, mode='rb') as file:
                fileContent = file.read()
                bigBlob += bytes(pBlobHeader)
                bigBlob += fileContent
        except:
            print("Failed to read file '{}'".format(filePath))
            result = False
            break

if result:
    try:
        with open(outputFilename, mode='wb') as file:
            file.write(bigBlob)
    except:
        print("Failed to write to file '{}'".format(outputFilename))
        result = False

if result:
    sys.exit(0)
else:
    sys.exit(-1)

File: sw/application/application-toolkit/file-packer/FilePacker.cpp
/*******************************************************************************
Copyright (C) Altera Corporation
 
This code and the related documents are Altera copyrighted materials and your
use of them is governed by the express license under which they were provided to
you ("License"). This code and the related documents are provided as is, with no
express or implied warranties other than those that are expressly stated in the
License.
*******************************************************************************/

#include <iostream>
#include <fstream>
#include <cstring>
#include <filesystem>
#include <vector>

// To convert the output of this app to an object for linking:
// $LD -r -b binary -o WebServerHome.o WebServerHome.bin

class FileBlobHeader
{
public:
    uint32_t _size;
    char _relativeFileName[128];
};

int main(int argumentCount, char* argumentArray[])
{
	if (argumentCount != 3)
	{
		std::cout << "Usage: " << argumentArray[0] << " <source-folder> <output-file>\n";
		return -1;
	}

	size_t bigBlobSize = 0;
	size_t fileBlobHeaderSize = sizeof(FileBlobHeader);
	std::string startFolderStr = argumentArray[1];
	std::string outputFilename = argumentArray[2];
	std::filesystem::path startFolder = std::filesystem::absolute(startFolderStr);
	std::filesystem::recursive_directory_iterator directoryIterator(argumentArray[1]);

	std::vector<std::filesystem::directory_entry> fileList;

	// Iterate all the files in the start folder
	for (auto& filePath : directoryIterator)
	{
		if (filePath.is_regular_file())
		{
			size_t blobSize = filePath.file_size() + fileBlobHeaderSize;
			bigBlobSize += blobSize;

			fileList.push_back(filePath);
		}
	}

	// Allocate a blob to contain all the files, with their headers
	std::vector<uint8_t> bigBlob(bigBlobSize);
	uint8_t* pWriteBuffer = bigBlob.data();

	bool result = true;
	size_t filenameSize = sizeof(FileBlobHeader::_relativeFileName);

	for (auto& filePath : fileList)
	{
		size_t fileSize = filePath.file_size();
		size_t blobSize = fileSize + fileBlobHeaderSize;

		FileBlobHeader* pBlobHeader = (FileBlobHeader*)pWriteBuffer;
		pBlobHeader->_size = (uint32_t)blobSize;
		memset(pBlobHeader->_relativeFileName, 0, filenameSize);

		std::filesystem::path fullPath = filePath.path();
		std::filesystem::path absolutePath = std::filesystem::absolute(fullPath);

		// Strip front of the path so it is relative to the start folder
		auto pathPartIt = absolutePath.begin();
		for (auto& startPathPart : startFolder)
			pathPartIt++;

		std::string relativePath;
		while (pathPartIt != absolutePath.end())
		{
			std::string part = pathPartIt->generic_u8string();
			pathPartIt++;

			if (pathPartIt != absolutePath.end())
				relativePath += (part + "/");
			else
				relativePath += part;
		}

		// Copy the relative path name into the header
		strncpy(pBlobHeader->_relativeFileName, relativePath.c_str(), filenameSize);

		// Read each file into the blob
		std::fstream inputStream;
		inputStream.open(filePath.path(), std::ios_base::in | std::ios_base::binary);
		if (!inputStream.good())
		{
			std::cout << "Failed to read file '" << filePath.path() << "'\n";
			result = false;
			break;
		}

		pWriteBuffer += fileBlobHeaderSize;
		inputStream.read((char*)pWriteBuffer, fileSize);
		pWriteBuffer += fileSize;
	}

	if (result)
	{
		std::fstream outputStream;
		outputStream.open(outputFilename.c_str(), std::ios_base::out | std::ios_base::binary);
		if (outputStream.good())
		{
			outputStream.write((char*)bigBlob.data(), bigBlob.size());
		}
		else
		{
			std::cout << "Failed to write to file '" << outputFilename << "'\n";
			result = false;
		}
	}

	return result ? 0 : -1;
}
